### PR TITLE
Comparison against TF-IDF Vectorizer (from scratch) #8

### PR DESCRIPTION
**Pull request by 2bit-hack**
## Related Issues & Documents (if needed)
Issue #8 

## What type of PR is this? (check all applicable)

- [ ] Refactoring existing code
- [X] Adding tests to existing code
- [X] New Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## QA Instructions(if necessary), Screenshots of Output or Recordings

`cd` into `tests` and run `tfidf_scratch_vs_rake.py`

## Added tests?

- [X] Yes
- [ ] No, because they aren't needed
- [ ] No, because I need help

## Added to documentation / readme or updated requirements.txt(if needed)?

- [ ] Changed documentation in readme
- [X] No documentation needed
- [ ] Updated requirements.txt

### CODE REVIEW

**Comment by 2bit-hack** [comment]
Should I incorporate Restyled's changes locally and make another PR?

**Comment by 2bit-hack** [comment]
Looks like Restyled automatically merged the changes into this PR

**Comment by 2bit-hack** [comment]
Okay, problem
Restyled reordered some of the imports so my fix of appending .. to the path isn't working

Right now, python can't get rake and tfidf imports from `tests`, because it can't import from a package above it
My fix makes the path search from `..tests`, that is, the root path, where it can find the imports
If Restyled changes the order of imports, the line that appends the .. happens after the imports and thus the imports fail

**Comment by BALaka-18** [comment]
> Okay, problem
> Restyled reordered some of the imports so my fix of appending .. to the path isn't working
> 
> Right now, python can't get rake and tfidf imports from `tests`, because it can't import from a package above it
> My fix makes the path search from `..tests`, that is, the root path, where it can find the imports
> If Restyled changes the order of imports, the line that appends the .. happens after the imports and thus the imports fail

But it shows all checks have passed. I don't think there's a problem :)

**Comment by 2bit-hack** [comment]
Try running the code and seeing if it works. If it works, that's great. On my system, I only managed to get this file working after that hacky fix of appending .., but I suppose it'll vary from system to system depending on how the Python path is set.

**Comment by 2bit-hack** [comment]
Does it import the rake and tfidf modules without any problem?

File: tfidf_vectorizer/extract_keywords_tfidf_scratch.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            else:
+                self.word_freq_docs[word] = 1
+
+        self.num_docs += 1
+
+    """
+    Builds a frequency counter of words from a corpus of docs
+    """
+
+    def digest_docs(self, docs):
```
Comment (BALaka-18) [inline tfidf_vectorizer/extract_keywords_tfidf_scratch.py]: @2bit-hack could you modify the name of this function ? It is too similar to the above one, and would create confusion, even though you've described it.
What about something like ```word_freq_docs``` ? Or something different from ```digest_doc```.
And also update the name of this function wherever you have used it next.

File: tests/tfidf_scratch_vs_rake.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+tfidf = TF_IDF_Scratch()
+tfidf_res = set([word[0] for word in tfidf.get_keywords(docs, 2)])
+end_tfidf = timer()
+
+start_rake = timer()
+rake = Rake()
+rake.get_keywords_from_raw_text(docs[2])
+rake_res = rake.get_ranked_keywords()
+end_rake = timer()
+
```
Comment (BALaka-18) [inline tests/tfidf_scratch_vs_rake.py:57]: Please use comments to state which function you're testing.

File: tests/tfidf_scratch_vs_rake.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+tfidf_tp = 0
+rake_total = len(rake_res)
+rake_tp = 0
+for keyphrase in manual_keyphrases:
+    if keyphrase in tfidf_res:
+        tfidf_tp += 1
+    if keyphrase in rake_res:
+        rake_tp += 1
+
+tfidf_precision = tfidf_tp / tfidf_total * 100
```
Comment (BALaka-18) [inline tests/tfidf_scratch_vs_rake.py:89]: Use comments to explain how and why you've used this formula.

File: tests/tfidf_scratch_vs_rake.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+end_rake = timer()
+
+print("tf-idf (scratch) results: ")
+print(tfidf_res)
+print()
+print("rake results: ")
+print(rake_res)
+print()
+
+tfidf_total = len(tfidf_res)
```
Comment (BALaka-18) [inline tests/tfidf_scratch_vs_rake.py:71]: Mention / explain using comments what this and the following lines are doing.

**Review by BALaka-18 — CHANGES_REQUESTED** [review]

**Comment by BALaka-18** [comment]
> Does it import the rake and tfidf modules without any problem?

Yes. Make the small changes and it'll be good to merge.
Also, squash all commits into a single commit after all commits have been made.

File: tfidf_vectorizer/extract_keywords_tfidf_scratch.py (hunk 1 above)
Comment (2bit-hack) [reply tfidf_vectorizer/extract_keywords_tfidf_scratch.py]: How about `build_word_freq_docs`?

**Comment by BALaka-18** [comment]
@2bit-hack it still shows 12 commits.

**Comment by 2bit-hack** [comment]
I'm not quite sure how to squash commits. I ran `git rebase` with the last 12 commits or so and squashed them into [one commit](https://github.com/BALaka-18/rake_new2/pull/20/commits/c53be13635bd53f3c5d3b63cf0f5c31c0cd31744) but it seems like the older commits haven't been removed.

**Comment by BALaka-18** [comment]
Okay not a problem I'll squash and merge for now.
But you might want to check this video out : https://www.youtube.com/watch?v=2E23I9PzplM

**Pull request merged**